jax/experimental/rnn.py

```diff
@@ -322,6 +334,22 @@ def lstm_cell(carry, x, *, W_ih, W_hh, b_ih, b_hh):
   c_n = jnp.stack(final_c)
   return seq_first_y.transpose(1, 0, 2), h_n, c_n
 
+def _extract_output(seq_lengths: Array, out):
+  _, ((hs, cs), seq_first_y) = out
+  h_t = _select_last(seq_lengths, hs)
+  c_t = _select_last(seq_lengths, cs)
+  # [seq_len, batch]       [1, batch]                               [seq_len, 1]
+  mask = seq_lengths[None] > jnp.arange(seq_first_y.shape[0])[:, None]
+  # [batch, seq_len, hidden_size]
+  seq_first_y = jnp.where(
+      mask[..., None], # [seq_len, batch, 1]
+      seq_first_y,     # [seq_len, batch, hidden_size]
+      0)
+  return (h_t, c_t), seq_first_y
+
+def _select_last(seq_lengths: Array, sequence: Array):
+  last_idx = seq_lengths[None, :, None] - 1
+  return jnp.take_along_axis(sequence, last_idx, axis=0)
```

Comment:
If I understand this correctly, seq_lengths is [batch_size] and sequence is [max_seq_len, batch_size, feature_dim].

We want an output that is [batch_size, feature_dim].

I think therefore we just need to return:
```
sequence[seq_lengths - 1, jnp.arange(sequence.shape[1])]
```

(I verified this internally)

Reply:
Thanks for catching this @sharadmv! I forgot to squeeze the first dimension when porting this code from Flax. I've confirmed your code is equivalent to:

```python
  last_idx = seq_lengths[None, :, None] - 1
  return jnp.take_along_axis(sequence, last_idx, axis=0)[0]
```
Will be keeping yours as its more readable.